src/execution/operator/aggregate/physical_window.cpp

```diff
@@ -1483,34 +1483,203 @@ unique_ptr<GlobalSinkState> PhysicalWindow::GetGlobalSinkState(ClientContext &co
 	return make_unique<WindowGlobalSinkState>(*this, context);
 }
 
-class WindowMergeTask : public ExecutorTask {
+class WindowGlobalMergeState;
+
+class WindowLocalMergeState {
 public:
-	WindowMergeTask(shared_ptr<Event> event_p, ClientContext &context_p, WindowGlobalHashGroup &hash_group_p)
-	    : ExecutorTask(context_p), event(move(event_p)), hash_group(hash_group_p) {
+	WindowLocalMergeState() : merge_state(nullptr) {
+		finished = true;
 	}
 
-	TaskExecutionResult ExecuteTask(TaskExecutionMode mode) override {
-		// Initialize merge sorted and iterate until done
-		auto &global_sort = *hash_group.global_sort;
-		MergeSorter merge_sorter(global_sort, global_sort.buffer_manager);
-		merge_sorter.PerformInMergeRound();
-		event->FinishTask();
-		return TaskExecutionResult::TASK_FINISHED;
+	bool TaskFinished() {
+		return finished;
+	}
+	void ExecuteTask();
+
+	WindowGlobalMergeState *merge_state;
+	atomic<bool> finished;
+};
+
+class WindowGlobalMergeState {
+public:
+	enum class SortStage : uint8_t { PREPARE, MERGE, SORTED };
+	explicit WindowGlobalMergeState(GlobalSortState &sort_state) : sort_state(sort_state), stage(SortStage::PREPARE) {
+		total_tasks = 0;
+		tasks_assigned = 0;
+		tasks_completed = 0;
+	}
+
+	bool IsSorted() const {
+		return stage == SortStage::SORTED;
+	}
+
+	bool AssignTask(WindowLocalMergeState &local_state);
+	bool TryPrepareNextRound(atomic<idx_t> &sorts_completed);
+
+	GlobalSortState &sort_state;
+
+	mutex lock;
+	SortStage stage;
+	atomic<idx_t> total_tasks;
+	atomic<idx_t> tasks_assigned;
+	atomic<idx_t> tasks_completed;
+};
+
+void WindowLocalMergeState::ExecuteTask() {
+	auto &global_sort = merge_state->sort_state;
+	MergeSorter merge_sorter(global_sort, global_sort.buffer_manager);
+	merge_sorter.PerformInMergeRound();
+	merge_state->tasks_completed++;
+	finished = true;
+}
+
+bool WindowGlobalMergeState::AssignTask(WindowLocalMergeState &local_state) {
+	lock_guard<mutex> guard(lock);
+
+	if (tasks_assigned >= total_tasks) {
+		return false;
 	}
 
+	local_state.merge_state = this;
+	local_state.finished = false;
+	tasks_assigned++;
+
+	return true;
+}
+
+bool WindowGlobalMergeState::TryPrepareNextRound(atomic<idx_t> &sorts_completed) {
+	lock_guard<mutex> guard(lock);
+
+	if (tasks_completed < total_tasks) {
+		return false;
+	}
+
+	tasks_assigned = tasks_completed = 0;
+
+	switch (stage) {
+	case SortStage::PREPARE:
+		sort_state.PrepareMergePhase();
```

Comment:
Are all threads busy during this `Event`?

I am a bit concerned about calling `PrepareMergePhase` while holding the lock. Other threads may have to wait until the thread holding the lock has completed `PrepareMergePhase`, even though they could be preparing the merge phase for other sorts.

Would it be possible to assign a `PREPARE` task to a thread, then drop the lock, and then the thread can do the preparation lock-free? If all threads are busy and this is not a problem, feel free to ignore this.

As a side note, I am looking at the code for `PrepareMergePhase`. This re-evaluates whether we need to be doing an external sort, by checking whether the variable-size data fits in memory. If it does, it unswizzles pointers so that the variable-size data does not have to be moved during the merge sort. However, if `external` is already `true` at this time, no unswizzling happens, and `PrepareMergePhase` is not costly at all. This check is not accurate for the window operator, because you have more than one sort going on at a time.

Not something you have to fix in this PR, but something to keep in mind.

Reply:
Yes I'm concerned about the relatively high levels of lock contention I was observing, especially near the end (the 'find a hash group" loop is N^2 in the numbers of groups.). 

Which operations are slow? I somehow got the impression that they were all fast except the merging, but now that I think about it, I got started down this road by noticing that unswizzling during prepare was time-consuming.

Reply:
I think the pattern here is to only grab the lock to assign a task / mark a task as done, and perform all other work lock-free. However, we should look into why PrepareMergePhase is always unswizzling. This should only happen if the data is close to the memory limit.

Reply:
Yeah the new PR does this.  And my test is rather brutal (Fannie Mae SF 4 uses about 70GB)